Reconstruction/RecCalorimeter/src/components/NoiseCaloCellsFromFileTool.cpp

```diff
@@ -0,0 +1,206 @@
+  declareProperty("readoutName", m_readoutName="ECalHitsPhiEta");
+  declareProperty("activeFieldName", m_activeFieldName="active_layer");
+  declareProperty("elecNoiseHistoName", m_elecNoiseHistoName="h_elecNoise_layer");
+  declareProperty("pileupHistoName", m_pileupHistoName="h_pileup_layer");
+  declareProperty("numRadialLayers",m_numRadialLayers=3);
+  //remove cells with energy bellow filterThreshold (threshold is multiplied by a cell noise sigma)
+  declareProperty("filterNoiseThreshold", m_filterThreshold=3);
+}
+
+NoiseCaloCellsFromFileTool::~NoiseCaloCellsFromFileTool() {}
+
+StatusCode NoiseCaloCellsFromFileTool::initialize() {
+  // Get GeoSvc
+  m_geoSvc = service ("GeoSvc");
+  if (!m_geoSvc) {
+    error() << "Unable to locate Geometry Service. "
+            << "Make sure you have GeoSvc and SimSvc in the right order in the configuration." << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  //Initialize random service
+  if(service( "RndmGenSvc" , m_randSvc ).isFailure()) {
+    error() << "Couldn't get RndmGenSvc!!!!" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  m_gauss.initialize(m_randSvc, Rndm::Gauss(0.,1.));
+
+  //open and check file, read the histograms with noise constants
+  if (initNoiseFromFile().isFailure()) {
+    error() << "Couldn't open file with noise constants!!!" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  // Get PhiEta segmentation
+  m_segmentation = dynamic_cast<DD4hep::DDSegmentation::GridPhiEta*>(
+    m_geoSvc->lcdd()->readout(m_readoutName).segmentation().segmentation());
+  if(m_segmentation == nullptr) {
+    error()<<"There is no phi-eta segmentation."<<endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  debug() << "Filter noise threshold: " <<  m_filterThreshold << "*sigma" << endmsg;
+
+  StatusCode sc = GaudiTool::initialize();
+  if (sc.isFailure()) return sc;
+
+  return sc;
+
+}
+
+void NoiseCaloCellsFromFileTool::addRandomCellNoise(std::unordered_map<uint64_t, double>& aCells) {
+  std::for_each( aCells.begin(), aCells.end(),
+    [this](std::pair<const uint64_t , double>& p) {p.second += ( getNoiseConstantPerCell(p.second) * m_gauss.shoot() );} );
+}
+
+void NoiseCaloCellsFromFileTool::filterCellNoise(std::unordered_map<uint64_t, double>& aCells) {
+  //Erase a cell if it has energy bellow a threshold from the vector
+  auto it = aCells.begin();
+  while ((it = std::find_if(it, aCells.end(),
+        [this](std::pair<const uint64_t,double>& p){
+          return bool(p.second < m_filterThreshold * getNoiseConstantPerCell(p.first));}
+        )) != aCells.end()) {
+    aCells.erase(it++);
+  }
+}
+
+StatusCode NoiseCaloCellsFromFileTool::finalize() {
+  //Delete histograms & file
+  debug() << "Deleting histograms with noise constants" << endmsg;
+  for (auto it: m_histoElecNoiseConst) {
+    delete it;
+  }
+  if (m_addPileup) {
+    for (auto it: m_histoPileupConst) {
+      delete it;
+    }
+  }
+  delete m_file;
```

Comment:
This (and above) looks like smart-pointers could be sensible here. Or no pointers at all.